### ISSUE TITLE
Recommendations on joining variants across callers for input to varlociraptor

### ISSUE THREAD

**Issue by nh13** [varlociraptor/varlociraptor#149]
The [calling docs](https://varlociraptor.github.io/docs/calling/)
>  In particular, you can also first combine candidates from different callers, to cover diverse variant types and length ranges

This is fairly intuitive if the candidates are _different_ variant types, however what if there are overlapping or duplicated variants?  An extreme example would be running multiple variant variant callers (eg. DeepVariant, GATK, etc.) to produce candidates and `varlociraptor` to produce a final set of calls.  In effect, we're using Varlociraptor to produce the final consensus calls among all these callers.

**Comment by dlaehnemann** [comment]
Varlociraptor will simply look at every entry in the `candidates.vcf` that you give it and evaluate the alternative haplotype(s) in comparison to the reference haplotype with its probabilistic model. So all it really needs are `CHROM`, `POS`, `REF` and `ALT` (unless you are looking at something more complicated, like breakends), and all the infos for the model are then gathered from the `BAM` file. Thus, if you have multiple entries (or rows or records) referring to the same genomic site, they will be evaluated independently. One exception is the handling of breakends, where varlociraptor simply collects them on a first pass over the `candidates.vcf` to jointly handle them afterwards---then, all breakends for a single event should be available.

So, if I understand your question correctly, I think this comes down to careful merging of different candidates generated (and the hassle this might involve with VCF/BCF files). I guess [`bcftools norm`](https://www.htslib.org/doc/bcftools.html#norm) and [the preprocessing capabilities of `hap.py`, `prep.py`,](https://github.com/Illumina/hap.py/blob/master/doc/normalisation.md) would be good, to initially left-align all indel records and do other normalization of representations, before creating proper multiallelic records.

A good standard on how to do get a most comprehensive and coherent set of candidates calls is definitely something worth exploring in more detail. Which tools to use, to get the greatest sensitivity. And how to merge them, to remove as much redundancy as possible, without creating misrepresentations of variants. Any input and discussion on this is welcome!

**Comment by johanneskoester** [comment]
Yeah, merging of candidates from overlapping callers is still a quite open question. Indeed, bcftools norm could be a solution, but one has to keep in mind that this will also destroy things like replacements as they are called by freebayes. Another option is to merge in impact space. The advantage is that all the genomic information is retained until the very end. So, one could just have a trivial merging of exactly equal variants after generating all the candidates (so that varlociraptor does not do any work twice). Then, run varlociraptor, and annotate e.g. with VEP. Afterwards, merge based on the effect on the protein. For those variants not hitting a protein, the problem remains though. There, left shifting of indels is a good option. The rest I'd probably leave as is.

**Comment by jdidion** [comment]
@nh13 where did you end up getting with this? I have a need to do something similar. I'm considering the best way to approach two separate issues:

1. Optimal settings for the individual variant callers to generate candidates (i.e. maximize recall). It would be nice to put together a list of recipes for popular callers.
2. Optimal method of merging candidate call sets. In the past I've used a combination of `vt` and `bcftools` for this, but it seems now that everything could be done with `bcftools norm -a -m -` and `bcftools concat -a -d`. (Of course this causes you to lose important information about complex variants so it's really only appropriate for benchmarking in the scenario where you're also normalizing the truth set in the same way.)

Another tool to investigate for (2) is [vcfwave](https://github.com/vcflib/vcflib/blob/master/doc/vcfwave.md). Or one could do a simple concatenation of the candidate VCF files and then post-process the varlociraptor output to select the best variant for each set of overlapping variants (but it would take some work to figure out heuristics for resolving complex cases).
